pkg/kubelet/apis/podresources/types.go

```diff
@@ -19,12 +19,15 @@ package podresources
 import (
 	"k8s.io/api/core/v1"
 	podresourcesapi "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"k8s.io/kubernetes/pkg/kubelet/cm/devicemanager"
 )
 
 // DevicesProvider knows how to provide the devices used by the given container
 type DevicesProvider interface {
 	GetDevices(podUID, containerName string) []*podresourcesapi.ContainerDevices
 	UpdateAllocatedDevices()
+	GetAllocatableDevices() devicemanager.ResourceDeviceInstances
```

Comment:
Hmm. I think you propagated this a bit too far. Here is where I would expect this now to be of type `podresourcesapi.ContainerDevices`. It is at this layer that the translation from `devicemanager.ResourceDeviceInstances` to ``podresourcesapi.ContainerDevices` should happen.

Reply:
This is a good point. The problem is - and here's why I didn't implement the change yet - is this layer is up until now extremely thin, like interface thin. We pass to the podresources server the concrete managers that implement the interface, without middleman translation layer. See: https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/kubelet.go#L2237 , where `containerManager` here is a https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/cm/container_manager.go#L44 .
So it takes a bit of rearchitecting. Let me think about it.

Reply:
OK, well maybe the right thing to do then, is to explicitly make the `containermanager` a `podresources.DevicesProvider` and a `podresources.CPUsProvider`. 

Then it wouldn't be strange for it to be returning these types.

Reply:
i.e., make the `contaienrmanager` interface:
```
type ContainerManager interface {
	// Runs the container manager's housekeeping.
	// - Ensures that the Docker daemon is in a container.
	// - Creates the system container where all non-containerized processes run.
	Start(*v1.Node, ActivePodsFunc, config.SourcesReady, status.PodStatusProvider, internalapi.RuntimeService) error
	// SystemCgroupsLimit returns resources allocated to system cgroups in the machine.
	// These cgroups include the system and Kubernetes services.
	SystemCgroupsLimit() v1.ResourceList
	// GetNodeConfig returns a NodeConfig that is being used by the container manager.
	GetNodeConfig() NodeConfig
	// Status returns internal Status.
	Status() Status
	// NewPodContainerManager is a factory method which returns a podContainerManager object
	// Returns a noop implementation if qos cgroup hierarchy is not enabled
	NewPodContainerManager() PodContainerManager
	// GetMountedSubsystems returns the mounted cgroup subsystems on the node
	GetMountedSubsystems() *CgroupSubsystems
	// GetQOSContainersInfo returns the names of top level QoS containers
	GetQOSContainersInfo() QOSContainersInfo
	// GetNodeAllocatableReservation returns the amount of compute resources that have to be reserved from scheduling.
	GetNodeAllocatableReservation() v1.ResourceList
	// GetCapacity returns the amount of compute resources tracked by container manager available on the node.
	GetCapacity() v1.ResourceList
	// GetDevicePluginResourceCapacity returns the node capacity (amount of total device plugin resources),
	// node allocatable (amount of total healthy resources reported by device plugin),
	// and inactive device plugin resources previously registered on the node.
	GetDevicePluginResourceCapacity() (v1.ResourceList, v1.ResourceList, []string)
	// UpdateQOSCgroups performs housekeeping updates to ensure that the top
	// level QoS containers have their desired state in a thread-safe way
	UpdateQOSCgroups() error
	// GetResources returns RunContainerOptions with devices, mounts, and env fields populated for
	// extended resources required by container.
	GetResources(pod *v1.Pod, container *v1.Container) (*kubecontainer.RunContainerOptions, error)
	// UpdatePluginResources calls Allocate of device plugin handler for potential
	// requests for device plugin resources, and returns an error if fails.
	// Otherwise, it updates allocatableResource in nodeInfo if necessary,
	// to make sure it is at least equal to the pod's requested capacity for
	// any registered device plugin resource
	UpdatePluginResources(*schedulerframework.NodeInfo, *lifecycle.PodAdmitAttributes) error
	InternalContainerLifecycle() InternalContainerLifecycle
	// GetPodCgroupRoot returns the cgroup which contains all pods.
	GetPodCgroupRoot() string
	// GetPluginRegistrationHandler returns a plugin registration handler
	// The pluginwatcher's Handlers allow to have a single module for handling
	// registration.
	GetPluginRegistrationHandler() cache.PluginHandler
	// ShouldResetExtendedResourceCapacity returns whether or not the extended resources should be zeroed,
	// due to node recreation.
	ShouldResetExtendedResourceCapacity() bool
	// GetAllocateResourcesPodAdmitHandler returns an instance of a PodAdmitHandler responsible for allocating pod resources.
	GetAllocateResourcesPodAdmitHandler() lifecycle.PodAdmitHandler
	// UpdateAllocatedDevices frees any Devices that are bound to terminated pods.
	UpdateAllocatedDevices()
        // Implements the podresources Provider API for CPUs and Devices
	podresources.CPUsProvider
	podresources.DevicesProvider
```

Reply:
Which is actually very similar to what you had originally (sorry for the churn), but now it is much more explicit.

Reply:
Nicer indeed! At first it felt to me a little weird to push `UpdateAllocatedDevices` into the `podresources.DevicesProvider` interface, but on second thought it's OK: the only other big consumer is `devicemanager` itself, and `ContainerManager` is just exposing this API to enable the `podresources` server. So it's now clearer that `ContainerManager` is just a delegating shim here.